features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/ResourceResolver.kt

```diff
@@ -0,0 +1,284 @@
+import java.util.concurrent.LinkedBlockingDeque
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+@Suppress("TooManyFunctions")
+internal class ResourceResolver(
+    private val bitmapCachesManager: BitmapCachesManager,
+    internal val threadPoolExecutor: ExecutorService = THREADPOOL_EXECUTOR,
+    private val drawableUtils: DrawableUtils,
+    private val webPImageCompression: ImageCompression,
+    private val logger: InternalLogger,
+    private val md5HashGenerator: MD5HashGenerator,
+    private val recordedDataQueueHandler: DataQueueHandler,
+    private val applicationId: String,
+    private val resourceItemCreationHandler: ResourceItemCreationHandler = ResourceItemCreationHandler(
+        recordedDataQueueHandler = recordedDataQueueHandler,
+        applicationId = applicationId
+    )
+) {
+
+    // region internal
+    @MainThread
+    internal fun resolveResourceId(
+        resources: Resources,
+        applicationContext: Context,
+        displayMetrics: DisplayMetrics,
+        drawable: Drawable,
+        drawableWidth: Int,
+        drawableHeight: Int,
+        resourcesSerializerCallback: ResourcesSerializerCallback
+    ) {
+        bitmapCachesManager.registerCallbacks(applicationContext)
+
+        val resourceId = tryToGetResourceFromCache(drawable = drawable)
+
+        if (resourceId != null) {
+            // if we got here it means we saw the bitmap before,
+            // so we don't need to send the resource again
+            resourcesSerializerCallback.onSuccess(resourceId)
+            return
+        }
+
+        val bitmapFromDrawable = if (drawable is BitmapDrawable && shouldUseDrawableBitmap(drawable)) {
+            drawable.bitmap // cannot be null - we already checked in shouldUseDrawableBitmap
+        } else {
+            null
+        }
+
+        // do in the background
+        threadPoolExecutor.executeSafe("resolveResourceId", logger) {
+            @Suppress("ThreadSafety") // this runs inside an executor
+            createBitmapAsync(
+                resources = resources,
+                drawable = drawable,
+                drawableWidth = drawableWidth,
+                drawableHeight = drawableHeight,
+                displayMetrics = displayMetrics,
+                bitmapFromDrawable = bitmapFromDrawable,
+                resolveResourceCallback = object : ResolveResourceCallback {
+                    override fun onResolved(resourceId: String, byteArray: ByteArray) {
+                        resourceItemCreationHandler.queueItem(resourceId, byteArray)
+                        resourcesSerializerCallback.onSuccess(resourceId)
+                    }
+
+                    override fun onFailed() {
+                        resourcesSerializerCallback.onFailure()
+                    }
+                }
+            )
+        }
+    }
+
+    // endregion
+
+    // region private
+
+    @WorkerThread
+    private fun createBitmapAsync(
```

Comment:
the method itself doesn't switch threads under the hood, so I guess we can remove `Async` from the name, because it is executed in the synchronous manner.